### DISCUSSION TITLE
API to create an in-process Hardhat network from a standalone program

### DISCUSSION THREAD

**Discussion by abacabadabacaba** [NomicFoundation/hardhat#5128] [Troubleshooting] [answered]
Using ganache, it is easy to create an ephemeral network from JavaScript:
```
const web3 = new Web3(require("ganache").provider(options));
// Now web3 is connected to an in-process network and can be used to interact with it.
```
How to do the same using hardhat-network? Hardhat will do it for you when running scripts in a Hardhat project, but I want to use it in a standalone program, not part of a project and run directly using `nodejs`.

**Comment by kanej** [comment, marked as answer]
There is not an easy way to create an ephemeral network based on a passed in config. You can import `hardhat` in a standalone script that will run a node:

```js
const hre = require("hardhat");
```

However this reads the config from a Hardhat config file in the current working directory - not what you are after. It is also limited to one Hardhat network per node process.

See the [docs for more details on use a Hardhat node from a script](https://hardhat.org/hardhat-runner/docs/advanced/scripts#standalone-scripts:-using-hardhat-as-a-library).

This is a limitation we explicitly looking to overcome in Hardhat v3, which we are actively working on.

**Comment by fer662** [comment]
@kanej since there's a 3.0.0 alpha as of now, did you ever get to work on this feature? If so, could you point me to the right spot in the alpha docs? Thanks!

**Reply by kanej** [reply]
Yes! Take a look at the Alpha docs: https://hardhat.org/hardhat3-alpha
For the moment they aren't at the level of detail you will need.
In the current Hardhat 3 Alpha (`v3.0.0-next.1`) you can do the following:

```ts
// ./scripts/create-hre-from-config.ts
import { createHardhatRuntimeEnvironment } from "hardhat/hre";
import { HardhatUserConfig } from "hardhat/types/config";

const config: HardhatUserConfig = {
  networks: {
    hardhat: {
      type: "edr",
      chainId: 7777,
    },
  },
};

const hre = await createHardhatRuntimeEnvironment(config);
const connection = await hre.network.connect();

const chainIdHex = (await connection.provider.request({
  method: "eth_chainId",
})) as string;

const chainId = parseInt(chainIdHex).toString();

console.log(`Connected to chain ID: ${chainId}`);
```
Which should give you something like:
```sh
> npx hardhat run ./scripts/create-hre-from-config.ts
# Compiling your Solidity contracts
# Compiled 1 Solidity file with solc 0.8.28 (evm target: cancun)

# Connected to chain ID: 7777
```
